File: software/examples/python/example_write_ndef.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

HOST = "localhost"
PORT = 4223
UID = "XYZ" # Change XYZ to the UID of your NFC/RFID Bricklet

from tinkerforge.ip_connection import IPConnection
from tinkerforge.bricklet_nfc import BrickletNFC

NDEF_URI = "www.tinkerforge.com"

# Callback function for state changed callback
def cb_state_changed(state, idle, nfc):
    if state == nfc.CARDEMU_STATE_IDLE:
        payload_uri = []

        for c in list(NDEF_URI):
            payload_uri.append(ord(c))

        # Only short records are supported.
        ndef_record_uri = [
                                0xD1,                 # MB/ME/CF/SR=1/IL/TNF
                                0x01,                 # TYPE LENGTH
                                len(payload_uri) + 1, # Length
                                ord("U"),             # Type
                                4                     # Status
                           ]

        for d in payload_uri:
            ndef_record_uri.append(d)

        nfc.cardemu_write_ndef(ndef_record_uri)
        nfc.cardemu_start_discovery()
    elif state == nfc.CARDEMU_STATE_DISCOVER_READY:
        nfc.cardemu_start_transfer(True)
    elif state == nfc.CARDEMU_STATE_DISCOVER_ERROR:
        print "Discover error"
    elif state == nfc.CARDEMU_STATE_TRANSFER_NDEF_ERROR:
        print "Transfer NDEF error"

if __name__ == "__main__":
    ipcon = IPConnection() # Create IP connection
    nfc = BrickletNFC(UID, ipcon) # Create device object

    ipcon.connect(HOST, PORT) # Connect to brickd
    # Don't use device before ipcon is connected

    # Register state changed callback to function cb_state_changed
    nfc.register_callback(nfc.CALLBACK_CARDEMU_STATE_CHANGED,
                          lambda x, y: cb_state_changed(x, y, nfc))

    nfc.set_mode(nfc.MODE_CARDEMU)

    raw_input("Press key to exit\n") # Use input() in Python 3
    ipcon.disconnect()

File: software/examples/java/ExampleWriteNDEF.java
import com.tinkerforge.IPConnection;
import com.tinkerforge.BrickletNFC;

public class ExampleWriteNDEF {
	private static final String HOST = "localhost";
	private static final int PORT = 4223;

	// Change XYZ to the UID of your NFC Bricklet
	private static final String UID = "XYZ";
	private static final String NDEF_URI = "www.tinkerforge.com";

	// Note: To make the example code cleaner we do not handle exceptions. Exceptions
	//       you might normally want to catch are described in the documentation
	public static void main(String args[]) throws Exception {
		IPConnection ipcon = new IPConnection(); // Create IP connection
		BrickletNFC nfc = new BrickletNFC(UID, ipcon); // Create device object

		ipcon.connect(HOST, PORT); // Connect to brickd
		// Don't use device before ipcon is connected

		// Add state changed listener
		nfc.addCardemuStateChangedListener(new BrickletNFC.CardemuStateChangedListener() {
			public void cardemuStateChanged(int state, boolean idle) {
				if (state == BrickletNFC.CARDEMU_STATE_IDLE) {
					try {
						int[] NDEFRecordURI = new int[NDEF_URI.toCharArray().length + 5];

						// Only short records are supported
						NDEFRecordURI[0] = 0xD1;                              // MB/ME/CF/SR=1/IL/TNF
						NDEFRecordURI[1] = 0x01;                              // TYPE LENGTH
						NDEFRecordURI[2] = NDEF_URI.toCharArray().length + 1; // Length
						NDEFRecordURI[3] = (int)'U';                          // Type
						NDEFRecordURI[4] = 0x04;                              // Status

						for (int i = 0; i < NDEF_URI.toCharArray().length; i++) {
							NDEFRecordURI[5 + i] = (int)NDEF_URI.toCharArray()[i];
						}

						nfc.cardemuWriteNdef(NDEFRecordURI);
						nfc.cardemuStartDiscovery();
					}
					catch (Exception e) {
						return;
					}
				}
				else if(state == BrickletNFC.CARDEMU_STATE_DISCOVER_READY) {
					try{
						nfc.cardemuStartTransfer(1);
					}
					catch (Exception e) {
						return;
					}
				}
				else if (state == BrickletNFC.CARDEMU_STATE_DISCOVER_ERROR) {
					System.out.println("Discover error");
				}
				else if (state == BrickletNFC.CARDEMU_STATE_TRANSFER_NDEF_ERROR) {
					System.out.println("Transfer NDEF error");
				}
				else if (state == BrickletNFC.READER_STATE_REQUEST_TAG_ID_ERROR) {
					System.out.println("Request tag ID error");
				}
			}
		});

		nfc.setMode(BrickletNFC.MODE_CARDEMU);

		System.out.println("Press key to exit"); System.in.read();
		ipcon.disconnect();
	}
}
